File: 01FlutterApps/middle_layer_springboot/knpiano_java/src/main/java/com/liu/springboot04web/bean/Kn01L002LsnBean.java
package com.liu.springboot04web.bean;

import java.time.LocalDate;
import java.time.ZoneId;
import java.util.Date;

import org.springframework.format.annotation.DateTimeFormat;

import com.fasterxml.jackson.annotation.JsonFormat;

public class Kn01L002LsnBean implements KnPianoBean {

    protected String lessonId;
    protected String subjectId;
    protected String subjectSubId;
    protected String subjectName;
    protected String subjectSubName;
    protected String stuId;
    protected String stuName;
    protected String nikName;
    protected Integer classDuration;
    protected Integer lessonType;
    protected Integer schedualType;
    // @JsonFormat(pattern = "yyyy-MM-dd HH:mm", 
    // //  timezone = "GMT+9" // 采用东京标准时区，接受手机前端的请求时接纳前端String类型的日期值
    // timezone = "GMT+8" // 采用新加坡标准时区，接受手机前端的请求时接纳前端String类型的日期值
    // )
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm") // timezone = "GMT+8" 的设置被统一到了application.properties里
    @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm")// 后台维护页面的请求响应处理
    protected Date schedualDate;

    // @JsonFormat(pattern = "yyyy-MM-dd HH:mm", 
    // //  timezone = "GMT+9" // 采用东京标准时区，接受手机前端的请求时接纳前端String类型的日期值
    // timezone = "GMT+8" // 采用新加坡标准时区，接受手机前端的请求时接纳前端String类型的日期值
    // )
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm") // timezone = "GMT+8" 的设置被统一到了application.properties里
    @DateTimeFormat(pattern = "yyyy-MM-dd")
    protected Date scanQrDate;

    // @JsonFormat(pattern = "yyyy-MM-dd HH:mm", 
    // //  timezone = "GMT+9" // 采用东京标准时区，接受手机前端的请求时接纳前端String类型的日期值
    // timezone = "GMT+8" // 采用新加坡标准时区，接受手机前端的请求时接纳前端String类型的日期值
    // )
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm") // timezone = "GMT+8" 的设置被统一到了application.properties里
    @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm")
    protected Date lsnAdjustedDate;

    // @JsonFormat(pattern = "yyyy-MM-dd HH:mm", 
    // //  timezone = "GMT+9" // 采用东京标准时区，接受手机前端的请求时接纳前端String类型的日期值
    // timezone = "GMT+8" // 采用新加坡标准时区，接受手机前端的请求时接纳前端String类型的日期值
    // )
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm") // timezone = "GMT+8" 的设置被统一到了application.properties里
    @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm")
    protected Date extraToDurDate;

    // @JsonFormat(pattern = "yyyy-MM-dd HH:mm", 
    // //  timezone = "GMT+9" // 采用东京标准时区，接受手机前端的请求时接纳前端String类型的日期值
    // timezone = "GMT+8" // 采用新加坡标准时区，接受手机前端的请求时接纳前端String类型的日期值
    // )
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm") // timezone = "GMT+8" 的设置被统一到了application.properties里
    @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm")
    protected Date originalSchedualDate; // 加课换正课后记录原来实际的计划上课日期
    protected String memo;
    protected Integer delFlg;
    protected Integer isFromPiceseLsn; // 是不是用零碎加课拼凑的课[0:不是拼凑的课。 1:是拼凑的课]
    protected Integer isExtraToScheLsn; // 是不是用加课换成的正课[0:不是。 1:是]
    protected Date createDate;
    protected Date updateDate;
// 变更，删除，签到，撤销，四个按钮在画面上活性/非活性的状态设置
    protected boolean usableEdit;
    protected boolean usableDelete;
    protected boolean usableSign;
    protected boolean usableCancel;
    protected boolean isToday;
// 其他
    protected Integer payStyle;
    protected String lsnMonth;    //监视排课异常数据模块用
    protected float subjectPrice; //监视排课异常数据模块用

    public float getSubjectPrice() {
        return subjectPrice;
    }
    public void setSubjectPrice(float subjectPrice) {
        this.subjectPrice = subjectPrice;
    }
    public String getLsnMonth() {
        return lsnMonth;
    }
    public void setLsnMonth(String lsnMonth) {
        this.lsnMonth = lsnMonth;
    }
    public String getLessonId() {
        return lessonId;
    }
    public void setLessonId(String lessonId) {
        this.lessonId = lessonId;
    }
    public String getSubjectId() {
        return subjectId;
    }
    public void setSubjectId(String subjectId) {
        this.subjectId = subjectId;
    }
    public String getSubjectName() {
        return subjectName;
    }
    public void setSubjectName(String subjectName) {
        this.subjectName = subjectName;
    }
    public String getStuId() {
        return stuId;
    }
    public void setStuId(String stuId) {
        this.stuId = stuId;
    }
    public String getStuName() {
        return stuName;
    }
    public void setStuName(String stuName) {
        this.stuName = stuName;
    }
        public String getNikName() {
        return nikName;
    }
    public void setNikName(String nikName) {
        this.nikName = nikName;
    }
    public Integer getClassDuration() {
        return classDuration;
    }
    public void setClassDuration(Integer classDuration) {
        this.classDuration = classDuration;
    }
    public Integer getLessonType() {
        return lessonType;
    }
    public void setLessonType(Integer lessonType) {
        this.lessonType = lessonType;
    }
    public Date getSchedualDate() {
        return schedualDate;
    }
    public void setSchedualDate(Date schedualDate) {
        this.schedualDate = schedualDate;
    }
    public Date getLsnAdjustedDate() {
        return lsnAdjustedDate;
    }
    public void setLsnAdjustedDate(Date lsnAdjustedDate) {
        this.lsnAdjustedDate = lsnAdjustedDate;
    }
    public Date getExtraToDurDate() {
        return extraToDurDate;
    }
    public void setExtraToDurDate(Date extraToDurDate) {
        this.extraToDurDate = extraToDurDate;
    }
    public Integer getDelFlg() {
        return delFlg;
    }
    public void setDelFlg(Integer delFlg) {
        this.delFlg = delFlg;
    }
    public Date getCreateDate() {
        return createDate;
    }
    public void setCreateDate(Date createDate) {
        this.createDate = createDate;
    }
    public Date getUpdateDate() {
        return updateDate;
    }
    public void setUpdateDate(Date updateDate) {
        this.updateDate = updateDate;
    }
// 变更，删除，签到，撤销，四个按钮在画面上活性/非活性的状态设置
    public void setUsableEdit(boolean usableEdit) {
        this.usableEdit = usableEdit;
    }
    public void setUsableDelete(boolean usableDelete) {
        this.usableDelete = usableDelete;
    }
    public void setUsableSign(boolean usableSign) {
        this.usableSign = usableSign;
    }
    public void setUsableCancel(boolean usableCancel) {
        this.usableCancel = usableCancel;
    }
    public boolean isUsableEdit() {
        return usableEdit;
    }
    public boolean isUsableDelete() {
        return usableDelete;
    }
    public boolean isUsableSign() {
        return usableSign;
    }
    public boolean isUsableCancel() {
        return usableCancel;
    }
    public boolean isToday(Date schedualDate) {
        // 获取当前日期
        LocalDate today = LocalDate.now();
        // 将传入的 Date 对象转换为 LocalDate 对象
        LocalDate schedualLocalDate = schedualDate.toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
        
        // 比较两个 LocalDate 是否相等
        return schedualLocalDate.equals(today);
    }
    public Date getScanQrDate() {
        return scanQrDate;
    }
    public void setScanQrDate(Date scanQrDate) {
        this.scanQrDate = scanQrDate;
    }
    public boolean isToday() {
        return isToday;
    }
    public void setToday(boolean isToday) {
        this.isToday = isToday;
    }
    public Integer getPayStyle() {
        return payStyle;
    }
    public void setPayStyle(Integer payStyle) {
        this.payStyle = payStyle;
    }
    public String getSubjectSubId() {
        return subjectSubId;
    }
    public void setSubjectSubId(String subjectSubId) {
        this.subjectSubId = subjectSubId;
    }
    public String getSubjectSubName() {
        return subjectSubName;
    }
    public void setSubjectSubName(String subjectSubName) {
        this.subjectSubName = subjectSubName;
    }
    public Integer getSchedualType() {
        return schedualType;
    }
    public void setSchedualType(Integer schedualType) {
        this.schedualType = schedualType;
    } 
    public Date getOriginalSchedualDate() {
        return originalSchedualDate;
    }
    public void setOriginalSchedualDate(Date originalSchedualDate) {
        this.originalSchedualDate = originalSchedualDate;
    }
    public String getMemo() {
        return memo;
    }
    public void setMemo(String memo) {
        this.memo = memo;
    }
    public Integer getIsFromPiceseLsn() {
        return isFromPiceseLsn;
    }
    public void setIsFromPiceseLsn(Integer isFromPiceseLsn) {
        this.isFromPiceseLsn = isFromPiceseLsn;
    }
    public Integer getIsExtraToScheLsn() {
        return isExtraToScheLsn;
    }
    public void setIsExtraToScheLsn(Integer isExtraToScheLsn) {
        this.isExtraToScheLsn = isExtraToScheLsn;
    }
}

File: 01FlutterApps/front_dart_flutter/lib/01LessonMngmnt/1LessonSchedual/Kn01L002LsnBean.dart
// ignore: file_names
import 'package:intl/intl.dart';

import '../../CommonProcess/CommonMethod.dart';

class Kn01L002LsnBean {
  final String lessonId;
  final String stuId;
  final String subjectId;
  final String subjectSubId;
  final String stuName;
  String? nikName; //因为从后台返回的nikName有可能是NULL
  final String subjectName;
  final String subjectSubName;
  late final int classDuration;
  final int lessonType;
  final String schedualDate; // 将其标记为 final，因为它在构造函数中被赋值。
  final String time;
  final String scanQrDate;
  final String lsnAdjustedDate;
  final String extraToDurDate;
  final String originalSchedualDate;
  String? memo; //为了可以NULL，不用final关键字
  final int isFromPiceseLsn; // 添加这个字段

  // 更新构造函数以正确赋值
  Kn01L002LsnBean({
    required this.lessonId,
    required this.stuId,
    required this.subjectId,
    required this.subjectSubId,
    required this.stuName,
    this.nikName,
    required this.subjectName,
    required this.subjectSubName,
    required this.classDuration,
    required this.lessonType,
    required this.schedualDate,
    required this.time,
    required this.scanQrDate,
    required this.lsnAdjustedDate,
    required this.extraToDurDate,
    required this.originalSchedualDate,
    this.memo, //为了可以NULL，不用required关键字
    required this.isFromPiceseLsn, // 添加这个参数
  });

  // 更新工厂方法以正确解析和格式化日期
  factory Kn01L002LsnBean.fromJson(Map<String, dynamic> json) {
    String formattedSchedualDate = '';
    String formattedTime = '';
    String formattedScanQrDate = "";
    String formattedLsnAdjustedDate = "";
    String formattedExtraToDurDate = "";
    String formattedOriginalSchedualDate = "";

    try {
      if (json['schedualDate'] != null && json['schedualDate'] != '') {
        DateTime parsedDate =
            CommonMethod.parseServerDate(json['schedualDate']);

        // 已经不再需要DateFormat('yyyy-MM-dd HH:mm').format(parsedDate.toLocal());的处理了
        // 因为时区已经在parseServerDate中处理
        formattedSchedualDate =
            DateFormat('yyyy-MM-dd HH:mm').format(parsedDate);
        formattedTime = DateFormat('HH:mm').format(parsedDate);
      }

      if (json['scanQrDate'] != null && json['scanQrDate'] != '') {
        DateTime parsedDate = CommonMethod.parseServerDate(json['scanQrDate']);
        formattedScanQrDate = DateFormat('yyyy-MM-dd').format(parsedDate);
      }

      if (json['lsnAdjustedDate'] != null && json['lsnAdjustedDate'] != '') {
        DateTime parsedDate =
            CommonMethod.parseServerDate(json['lsnAdjustedDate']);
        formattedLsnAdjustedDate =
            DateFormat('yyyy-MM-dd HH:mm').format(parsedDate);
        formattedTime = DateFormat('HH:mm').format(parsedDate);
      }

      if (json['extraToDurDate'] != null && json['extraToDurDate'] != '') {
        DateTime parsedDate =
            CommonMethod.parseServerDate(json['extraToDurDate']);
        formattedExtraToDurDate = DateFormat('yyyy-MM-dd').format(parsedDate);
      }

      if (json['originalSchedualDate'] != null &&
          json['originalSchedualDate'] != '') {
        DateTime parsedDate =
            CommonMethod.parseServerDate(json['originalSchedualDate']);
        formattedOriginalSchedualDate =
            DateFormat('yyyy-MM-dd HH:mm').format(parsedDate);
        formattedTime = DateFormat('HH:mm').format(parsedDate);
      }
    } catch (e) {
      print('Error parsing or formatting schedualDate: $e');
    }

    return Kn01L002LsnBean(
      lessonId: json['lessonId'] as String,
      stuId: json['stuId'] as String,
      subjectId: json['subjectId'] as String,
      subjectSubId: json['subjectSubId'] as String,
      stuName: json['stuName'] as String,
      nikName: json['nikName'] as String?,
      subjectName: json['subjectName'] as String,
      subjectSubName: json['subjectSubName'] as String,
      classDuration: json['classDuration'] as int,
      lessonType: json['lessonType'] as int,
      schedualDate: formattedSchedualDate,
      time: formattedTime,
      scanQrDate: formattedScanQrDate,
      lsnAdjustedDate: formattedLsnAdjustedDate,
      extraToDurDate: formattedExtraToDurDate,
      originalSchedualDate: formattedOriginalSchedualDate,
      memo: json['memo'] as String?,
      // 关键：确保字段名与后端返回的JSON字段名完全一致
      isFromPiceseLsn: json['isFromPiceseLsn'] as int? ?? 0, // 添加默认值保护
    );
  }
}
